core/src/test/java12/com/linecorp/armeria/internal/common/Java12ContextAwareFutureTest.java

```diff
@@ -0,0 +1,182 @@
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.ArgumentsSource;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Captor;
+import org.mockito.Mock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.util.concurrent.MoreExecutors;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.util.SafeCloseable;
+import com.linecorp.armeria.common.util.SystemInfo;
+import com.linecorp.armeria.internal.common.ContextFutureCallbackArgumentsProvider.CallbackResult;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Appender;
+
+class Java12ContextAwareFutureTest {
+
+    // TODO(minwoox) Make an extension which a user can easily check the logs.
+    @Mock
+    private Appender<ILoggingEvent> appender;
+    @Captor
+    private ArgumentCaptor<ILoggingEvent> eventCaptor;
+
+    @BeforeAll
+    static void checkEnv() {
+        assumeThat(SystemInfo.javaVersion()).isGreaterThanOrEqualTo(12);
+    }
+
+    @BeforeEach
+    void setupLogger() {
+        ((ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME))
+                .addAppender(appender);
+    }
+
+    @AfterEach
+    void cleanupLogger() {
+        ((ch.qos.logback.classic.Logger) LoggerFactory.getLogger(Logger.ROOT_LOGGER_NAME))
+                .detachAppender(appender);
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(Java12CallbackArgumentsProvider.class)
+    void makeContextAwareCompletableFutureShouldPropagateContext(
+            BiConsumer<CompletableFuture<?>, CallbackResult> callback) {
+        final HttpRequest req = HttpRequest.of(HttpMethod.GET, "/");
+        final ServiceRequestContext ctx1 = ServiceRequestContext.builder(req).build();
+        try (SafeCloseable ignored = ctx1.push()) {
+            final CompletableFuture<Object> future = new CompletableFuture<>();
+            final CompletableFuture<Object> contextAwareFuture = ctx1.makeContextAware(future);
+            final CallbackResult callbackCalled = new CallbackResult();
+            callback.accept(contextAwareFuture, callbackCalled);
+
+            future.complete("");
+
+            assertThat(callbackCalled.called.get()).isTrue();
+            verify(appender, atLeast(0)).doAppend(eventCaptor.capture());
+            assertThat(eventCaptor.getAllValues()).anySatisfy(event -> {
+                assertThat(event.getLevel()).isEqualTo(Level.DEBUG);
+                assertThat(event.getMessage()).startsWith("Router created for");
+            });
```

Comment:
The `eventCaptor.getAllValues()` does not guarantee happen-before relationship. That means this assertion could be flaky.
```suggestion
            await().untilAsserted(() -> {
                assertThat(eventCaptor.getAllValues()).anySatisfy(event -> {
                    assertThat(event.getLevel()).isEqualTo(Level.DEBUG);
                    assertThat(event.getMessage()).startsWith("Router created for");
                });
            });
```